### ISSUE TITLE
Compilation error with icc 18.0.1

### ISSUE THREAD

**Issue by rmcgibbo** [psi4/psi4#923]
Compiling psi4 on -O3 with icc 18.0.1 seems to trigger a crash in the compiler.

The crash is not observed with the icc 17.0.2 with -O3, nor is it observed with icc 18.0.1 if I turn the optimization flag down to -O1.

```
/u/nyc/mcgibbon/gerrit/psi4-gerrit/psi4/src/psi4/dcft/dcft_gradient_UHF.cc(3903) (col. 72): internal error: 20000_6805

[ 85%] Building CXX object src/psi4/optking/CMakeFiles/optking.dir/opt_data_io.cc.o
compilation aborted for <prefix>/psi4/src/psi4/dcft/dcft_gradient_UHF.cc (code 4)
make[5]: *** [src/psi4/dcft/CMakeFiles/dcft.dir/dcft_gradient_UHF.cc.o] Error 4
make[4]: *** [src/psi4/dcft/CMakeFiles/dcft.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
```

With 18.0.1, I can fix the error by turning down the optimization flag for that file in particular by adding the following snippet to psi4/src/psi4/dcft/CMakeLists.txt:
```
if (CMAKE_CXX_COMPILER_ID STREQUAL "Intel" AND
    CMAKE_CXX_COMPILER_VERSION VERSION_LESS 18.1 AND
    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 18.0)

    message(WARNING "Intel C++ (ICC) 18.0 fails to compile dcft_gradient_UHF.cc with aggressive optimization flags")
    set_source_files_properties(dcft_gradient_UHF.cc PROPERTIES
        COMPILE_FLAGS -O1)
endif ()
```

Has anyone else tried using icc 18.0.1? Is there precedent for adding compiler- and file-specific flags in the CMakeLists? I'm not sure if this is suitable as a PR.

**Comment by loriab** [comment]
I think @jturney has built just fine with 18.0.0 – have you tried 18.0.1, Jet?

No problem with compiler- and file-specific flags. We've done this [before](https://github.com/psi4/psi4/blob/f5560a0fe5a88a64dd6c040339b9f682e277da61/psi4/src/psi4/detci/CMakeLists.txt#L3) though in the other direction. Might see if Jet sees a code solution first.

**Comment by jturney** [comment]
I'm hitting a similar error message with `dcft_gradient_UHF.cc` but it reports that the error is in `libmints/matrix.h`.  Reducing the optimization level was the only way I could get it to compile.  Even compiling with -O2 gives the error, -O1 is the only way to resolve it.

I see no issue with the compiler- and file-specific flags for this.

**Comment by andysim** [comment]
I don't have a chance to compile with icpc 18 right now, but we do have it
at NIH.  One way to narrow this problem down is to wrap half of the
matrix.h functions in preprocessor pragmas to kill optimization (
https://software.intel.com/en-us/node/524550), leaving the rest compiled as
normal with -O2 or higher.  By eliminating half of the functions'
optimization each time, it should be quite easy to figure out which
function is causing problems by bisection.

**Comment by edeprince3** [comment]
I can confirm this problem with icc 18  and dcft_gradient_UHF.cc.  I cheated and just commented out the problematic lines, though.

**Comment by jturney** [comment]
@edeprince3 Was it the same lines that @rmcgibbo reported? dcft_gradient_UHF.cc(3903)

**Comment by edeprince3** [comment]
Yes, dcft_gradient_UHF.cc, line 3903 and and also line 3998.

**Comment by dgasmith** [comment]
That loop looks about as simple as it gets. What kills me is the loop structure ~10 lines above is nearly identical.

Any error message on the crash?

**Comment by rmcgibbo** [comment]
> Any error message on the crash?

No.

**Comment by rmcgibbo** [comment]
And that error code (`20000_6805`) doesn't turn up anything on google either.

**Comment by edeprince3** [comment]
Same here, just 
/edfs/users/deprince/software/psi4/psi4/src/psi4/dcft/dcft_gradient_UHF.cc(3903) (col. 72): internal error: 20000_6805

**Comment by rmcgibbo** [comment]
I guess if anyone has an account for submitting bug reports to the icc team... ¯\_(ツ)_/¯

**Comment by hokru** [comment]
some out of out-of-memory problem perhaps? (real or intel internal). It is quite a long file.

**Comment by dgasmith** [comment]
With ICC 17 and CMake 3.5.1 I get the following error:
```
CMake Error at src/psi4/dcft/CMakeLists.txt:4 (if):
  if given arguments:

    "CMAKE_CXX_COMPILER_ID" "STREQUAL" "Intel" "AND" "CMAKE_CXX_COMPILER_VERSION" "VERSION_LESS" "18.1" "AND" "CMAKE_CXX_COMPILER_VERSION" "VERSION_GREATER_EQUAL" "18.0"

  Unknown arguments specified
```

From @loriab's guesses either `STREQUAL` should `MATCHES` or we need a parentheses around each part.

**Comment by loriab** [comment]
yeah, next person who hits this should imitate [this](https://github.com/psi4/psi4/blob/master/cmake/custom_cxxstandard.cmake#L8-L9).

**Comment by rmcgibbo** [comment]
Ouch, sorry @dgasmith. It looks like I used VERSION_GREATER_EQUAL which was added in https://cmake.org/cmake/help/v3.7/release/3.7.html

**Comment by loriab** [comment]
This is fixed at Intel 2018.0.3, so flag check relaxed in acd3cb62ae1e9fc67ab890d67c550dd72f71e3f1
